File: LeetCode/sort_colors.rb
#  Given an array with n objects colored red, white or blue, sort them so that objects of the same color are adjacent, with the colors in the order red, white and blue.
#
#  Here, we will use the integers 0, 1, and 2 to represent the color red, white, and blue respectively. 
#

colors = { 1 => "red", 2 => "white", 3 => "blue" }

array = (1..Random.rand(6..20)).map { |i|
    Random.rand(1..3)
}

puts array.inspect

small = -1
undecided = 0
large = array.length

while (undecided < large)
    case array[undecided]
    when 3
        large -= 1
        array[large], array[undecided] = array[undecided], array[large]
    when 2
        undecided += 1
    when 1
        small += 1
        array[small], array[undecided] = array[undecided], array[small]
        undecided += 1
    end
end

puts array.map {|x| colors[x] }.inspect

File: LeetCode/sort_colors.cs
using System;

class Program {
    private enum Colors {
            red = 0,
            white,
            blue
    }

    static void Flag(Colors[] c) {
        int beginning = -1;
        int end = c.Length;
        int middle = 0;
        while (middle < end) {
            switch (c[middle]) {
                case Colors.red:
                    beginning++;
                    Swap(ref c[beginning], ref c[middle]);
                    middle++;
                    break;
                case Colors.white:
                    middle++;
                    break;
                case Colors.blue:
                    end--;
                    Swap(ref c[middle], ref c[end]);
                    break;
                default:
                    throw new NotSupportedException();
            }
        }
    }

    static void Swap(ref Colors a, ref Colors b) {
        Colors c = a;
        a = b;
        b = c;
    }

    static void Main() {
        int n = 15;
        Colors[] colors = new Colors[n];
        Random rand = new Random();
        for (int i = 0; i < n; i++) {
            colors[i] = (Colors) rand.Next(0, 3);
        }

        Console.WriteLine(String.Join(", ", colors));
        Flag(colors);
        Console.WriteLine(String.Join(", ", colors));
    }
}
